app/code/Magento/Bundle/Test/Unit/Observer/AppendUpsellProductsObserverTest.php

```diff
@@ -0,0 +1,411 @@
+            ->willReturn($this->eventMock);
+
+        $this->eventMock
+            ->expects($this->once())
+            ->method('getProduct')
+            ->willReturn($this->productMock);
+
+        $this->bundleDataMock
+            ->expects($this->once())
+            ->method('getAllowedSelectionTypes')
+            ->willReturn($this->getAllowedSelectionTypes());
+
+        $this->productMock
+            ->expects($this->once())
+            ->method('getTypeId')
+            ->willReturn($typeId);
+
+        $this->eventMock
+            ->expects($this->once())
+            ->method('getCollection')
+            ->willReturn($this->collectionMock);
+
+        $this->eventMock
+            ->expects($this->once())
+            ->method('getLimit')
+            ->willReturn($limit);
+
+        $this->collectionMock
+            ->expects($this->exactly(2))
+            ->method('getItems')
+            ->willReturn($collectionItems);
+
+        $this->productMock
+            ->expects($this->once())
+            ->method('getId')
+            ->willReturn($productId);
+
+        $this->bundleSelectionMock
+            ->expects($this->once())
+            ->method('getParentIdsByChild')
+            ->willReturn($parentIds);
+
+        $this->productMock
+            ->expects($this->once())
+            ->method('getCollection')
+            ->willReturn($this->bundleCollectionMock);
+
+        $this->bundleCollectionMock
+            ->expects($this->once())
+            ->method('addAttributeToSelect')
+            ->willReturn($this->bundleCollectionMock);
+
+        $this->configMock
+            ->expects($this->once())
+            ->method('getProductAttributes')
+            ->willReturn($productAttributes);
+
+        $this->productVisibilityMock
+            ->expects($this->once())
+            ->method('getVisibleInCatalogIds')
+            ->willReturn($visibleInCatalogIds);
+
+        $this->bundleCollectionMock
+            ->expects($this->once())
+            ->method('setPageSize')
+            ->willReturn($this->bundleCollectionMock);
+
+        $this->bundleCollectionMock
+            ->expects($this->once())
+            ->method('addFieldToFilter')
+            ->willReturn($this->bundleCollectionMock);
+
+        $this->bundleCollectionMock
+            ->expects($this->once())
+            ->method('setFlag')
+            ->willReturn($this->bundleCollectionMock);
+
+        $this->collectionMock
+            ->expects($this->once())
+            ->method('setItems')
+            ->willReturn($collectionItems);
+
+        $this->observer->execute($this->observerMock);
+    }
+
+    /**
+     * Test observer when collection contains a parent product of a current product
+     */
+    public function testObserverWithoutBundleIds()
+    {
+        $collectionItems = [
+            1 => 1
+        ];
+        $parentIds = [1];
+        $limit = 2;
+        $productId = 2;
+        $typeId = 'simple';
```

Comment:
You can refer to const from Magento Core

Reply:
**done**